Adds range with step for number of users

jeppe-style opened the pull request:
as per title

Fixes (for number of users) https://github.com/benchflow/benchflow/issues/397

Currently supported for number of users but open for extension to other values

@VincenzoFerme please carefully have a look at `RangeWithStep` because I made some assumptions there when to say it is not a valid specification

jeppe-style committed:
Adds range with step for number of users

VincenzoFerme commented on benchflow-dsl/src/test/scala/cloud/benchflow/dsl/definition/configuration/goal/explorationspace/explorationvalues/IntValuesTest.scala:
```diff
@@ -50,4 +50,92 @@ class IntValuesTest extends JUnitSuite {
 
   }
 
```
Cover also `+0`, `*1`, `^0`, `^1`, and in general all the [Absorbing Elements](https://en.wikipedia.org/wiki/Absorbing_element), [Identity Elements](https://en.wikipedia.org/wiki/Identity_element), and [Inverse Elements](https://en.wikipedia.org/wiki/Inverse_element) of the Operators we Have. Also negative numbers for all the operations. 

I saw we just support integer specification, and for now I think it is fine to cover the most common cases.

jeppe-style replied:
I added now more tests for the cases you specified. I also added subtraction so that we also can handle decreasing sequences.

VincenzoFerme replied:
We are safe from https://en.wikipedia.org/wiki/Inverse_element#Real_numbers for now because we just support integer numbers as steps.

VincenzoFerme commented on benchflow-dsl/src/test/scala/cloud/benchflow/dsl/definition/configuration/goal/explorationspace/explorationvalues/IntValuesTest.scala:
```diff
@@ -50,4 +50,92 @@ class IntValuesTest extends JUnitSuite {
+
+    val terminationCriteria = rangeStepYaml.parseYaml.convertTo[Try[IntValues]]
+
+    Assert.assertTrue(terminationCriteria.isFailure)
```
We should also report back the type of problem

jeppe-style replied:
we do report type of failure but in the test I think it is sufficient for now to just check that it actually failed as expected.

VincenzoFerme commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/configuration/goal/explorationspace/explorationvalues/ExplorationValuesIntYamlProtocol.scala:
```diff
@@ -62,10 +72,48 @@ object ExplorationValuesIntYamlProtocol extends DefaultYamlProtocol {
-    (range.head to range(1)).toList
+    } else {
+
+      Success((range.head to range(1)).toList)
```
Is this the default for a range? `+1` as step? If yes, make it clear in a comment.

jeppe-style replied:
added

VincenzoFerme commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/configuration/goal/explorationspace/step/RangeWithStep.scala:
```diff
@@ -0,0 +1,56 @@
+
+    val list = generateList(List(range.head), step, range(1), func)
+
+    if (list.size <= 2) {
```
Validation should happen also before calling this function, so you can also report to the user what is the issue. Here it is fine to check the result you get is fine.

jeppe-style replied:
I check that the range has two elements in the list, that the step function is valid and that the step itself is an integer. Not sure what we can test beyond that before running. If you have any ideas that would be helpful.

VincenzoFerme replied:
Yes, everything that math says is not valid (see https://github.com/benchflow/benchflow/pull/465#discussion_r123855300)

VincenzoFerme commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/configuration/goal/explorationspace/explorationvalues/ExplorationValuesIntYamlProtocol.scala:
```diff
@@ -62,10 +72,48 @@ object ExplorationValuesIntYamlProtocol extends DefaultYamlProtocol {
+      val prefix = step.substring(0, 1)
+
+      val triedValue = Try(step.substring(1).toInt)
+
```
You also need to validate the validity of the entire step definition, as discussed in a subsequent comment on the tests. Do not call and wait for a fail of the `RangeWithStep` function if you already know something it is not valid.

jeppe-style replied:
related to next comment

VincenzoFerme commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/configuration/goal/explorationspace/step/RangeWithStep.scala:
```diff
@@ -0,0 +1,56 @@
+      // to avoid infinite recursion
+      list
+
+    } else if (list.size == 1 && num.gt(num.abs(num.minus(maxValue, next)), num.abs(num.minus(maxValue, list.head)))) {
```
This case also, it is fine to have it here for security reason, but can be detected before as discussed in a comment on the test section.

jeppe-style replied:
I don't see how that can be done.

VincenzoFerme replied:
See https://github.com/benchflow/benchflow/pull/465#discussion_r123855300

VincenzoFerme commented:
@jeppe-style I also marked https://github.com/benchflow/benchflow/issues/397 as fixed only for number of users, so that we track the need of extending it to other cases.

jeppe-style committed:
Adds additional tests for range with step

VincenzoFerme commented on benchflow-dsl/src/test/scala/cloud/benchflow/dsl/definition/configuration/goal/explorationspace/explorationvalues/IntValuesTest.scala:
```diff
@@ -90,6 +110,26 @@ class IntValuesTest extends JUnitSuite {
+
+    val terminationCriteria = rangeStepYaml.parseYaml.convertTo[Try[IntValues]]
+
+    Assert.assertTrue(terminationCriteria.isSuccess)
```
Why is this valid? The range should be `a < b` for this to be valid.

jeppe-style commented:
Discussion:
- [x] add the error messages for the errors detected during generation of the list (e.g. return `Try[List[Numeric]]`)
- [x] errors are defined by messages (e.g. no hierarchy necessary for now)
- [x] requirement of the range is that it goes from best expected performance to worse expected performance
- [x] number of users should always be greater than 0
- [x] add validation for math (e.g. identity etc.) before calling `RangeWithStep.stepList`


-  to provide flexibility we can in the future add exploration strategies to cover all cases (e.g. invert exploration (worst to best))
- in general there should be specific control for each type of exploration space dimension

jeppe-style committed:
Improves validation of exploration space number of users
Refactors IntValues to UserValues

VincenzoFerme commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/configuration/goal/explorationspace/workload/users/UserValuesYamlProtocol.scala:
```diff
@@ -52,46 +53,66 @@ object ExplorationValuesIntYamlProtocol extends DefaultYamlProtocol {
 
   private def getRangeList(range: List[Int]): Try[List[Int]] = {
 
     if (range.size != 2) {
```
The validation part looks the same as the `getStepList` function: https://github.com/benchflow/benchflow/pull/465/files/1c9cced7a66798daefdb836ab19fc561df17bb7f..561cf5d61e4b9bddd849242517c01b03d56da5c7#diff-5fbdece869476b79c16c33a628a1057dR108

jeppe-style replied:
It is similar but not exactly the same. I wanted it to be explicit therefore I did not put it into a common method.

VincenzoFerme commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/configuration/goal/explorationspace/workload/users/UserValuesYamlProtocol.scala:
```diff
@@ -111,10 +137,26 @@ object ExplorationValuesIntYamlProtocol extends DefaultYamlProtocol {
+      }
+
+    case "^" =>
+      if (step == 1) {
```
Also `^0` is not valid, because you keep getting `1`.

jeppe-style replied:
I check this already in the in the `getStepList` method. I remove the check from the `+` case to  make it consistent.

VincenzoFerme commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/configuration/goal/explorationspace/workload/users/UserValuesYamlProtocol.scala:
```diff
@@ -111,10 +137,26 @@ object ExplorationValuesIntYamlProtocol extends DefaultYamlProtocol {
+      }
+
+    case "*" =>
+      if (step == 1) {
```
Also `*0` is not valid because you are keep getting `0` (https://en.wikipedia.org/wiki/Absorbing_element). Place also a reference to the way you perform these checks.

jeppe-style replied:
I added a comment that it is checked prior to calling this function

VincenzoFerme commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/configuration/goal/explorationspace/workload/users/UserValuesYamlProtocol.scala:
```diff
@@ -111,10 +137,26 @@ object ExplorationValuesIntYamlProtocol extends DefaultYamlProtocol {
   }
 
   private def generateStepList(range: List[Int], step: Int, prefix: String): Try[List[Int]] = prefix match {
```
The checks, other than the comments, are fine. But we should check also the seed number (so in our case the first element in the range) for the same values, because that is used as a first argument in the operation.

jeppe-style replied:
I added the check for the `^` case. For other cases there is no impact.

VincenzoFerme reviewed (changes requested):
These are the changes that we still to apply before being able to merge the pull request. Including: https://github.com/benchflow/benchflow/pull/465#pullrequestreview-46255861

VincenzoFerme commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/configuration/goal/explorationspace/workload/users/UserValuesYamlProtocol.scala:
```diff
@@ -0,0 +1,162 @@
+
+  }
+
+  private def getRangeList(range: List[Int]): Try[List[Int]] = {
```
General checks goes in the superclass, type specific steps have to be here.

jeppe-style replied:
see https://github.com/benchflow/benchflow/pull/465#discussion_r124842226

VincenzoFerme commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/configuration/goal/explorationspace/workload/users/UserValuesYamlProtocol.scala:
```diff
@@ -0,0 +1,162 @@
+
+  private def getStepList(range: List[Int], step: String): Try[List[Int]] = {
+
+    if (range.size != 2) {
```
General checks goes in the superclass, type specific steps have to be here.

VincenzoFerme replied:
Moreover most of the checks are duplicated with https://github.com/benchflow/benchflow/pull/465/files#diff-5fbdece869476b79c16c33a628a1057dR95

jeppe-style replied:
see https://github.com/benchflow/benchflow/pull/465#discussion_r124842226

VincenzoFerme commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/configuration/goal/explorationspace/workload/users/UserValuesYamlProtocol.scala:
```diff
@@ -0,0 +1,162 @@
+    override def read(yaml: YamlValue): UserValues = unsupportedReadOperation
+
+  }
+
```
The following function should be enforced from the superclass, so that when implementing a new type, it is clear they have to be defined. E.g., define them `abstract`.

jeppe-style replied:
see https://github.com/benchflow/benchflow/pull/465#discussion_r124842226

VincenzoFerme commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/configuration/goal/explorationspace/workload/users/UserValuesYamlProtocol.scala:
```diff
@@ -0,0 +1,162 @@
+
+  private def assertValidValues(values: List[Int]): Try[List[Int]] = {
+
+    // check that all values are in increasing order
```
General checks goes in the superclass, type specific steps have to be here.

jeppe-style replied:
see https://github.com/benchflow/benchflow/pull/465#discussion_r124842226

VincenzoFerme commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/configuration/goal/explorationspace/workload/users/UserValuesYamlProtocol.scala:
```diff
@@ -0,0 +1,162 @@
+    }
+
+    if (!increasingOrder) {
+      Failure(new BenchFlowDeserializationException("values must be increasing", new Throwable))
```
When you report this `Failure` is it clear in which part of the DSL definition there was the issue?

jeppe-style replied:
It was not. I have fixed this now.

VincenzoFerme commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/configuration/goal/explorationspace/workload/users/UserValuesYamlProtocol.scala:
```diff
@@ -0,0 +1,162 @@
+    } else if (range.head <= 0) {
+      Failure(new BenchFlowDeserializationException("range must be > 0", new Throwable))
+    } else {
+      // IF only range then the step is 1
```
This is a default we implicitly assume, make it clear in the class definition as we did for the other default values.

jeppe-style replied:
It is not specific to this class but in general therefore I wouldn't put it in the class definition.

VincenzoFerme commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/configuration/goal/explorationspace/workload/users/UserValuesYamlProtocol.scala:
```diff
@@ -0,0 +1,162 @@
+
+  }
+
+  private def assertValidValues(values: List[Int]): Try[List[Int]] = {
```
You can define a `MonotonicPositiveInt` type to check for this, extending a `MonotonicInt` type that just check for https://github.com/benchflow/benchflow/pull/465/files#diff-5fbdece869476b79c16c33a628a1057dR79. Then `MonotonicInt` can extend `Int` (see https://github.com/benchflow/benchflow/pull/465/files#diff-5fbdece869476b79c16c33a628a1057dR118). This way you increase reusability, especially for these cases that are very common in our DSL.

jeppe-style replied:
see https://github.com/benchflow/benchflow/pull/465#discussion_r124842226

VincenzoFerme commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/configuration/goal/explorationspace/workload/users/UserValuesYamlProtocol.scala:
```diff
@@ -0,0 +1,162 @@
+      Failure(new BenchFlowDeserializationException("range must be > 0", new Throwable))
+    } else {
+
+      val prefix = step.substring(0, 1)
```
Also for this you can have a type, say `PositiveIntStep` extending `IntStep` extending a general type `Int`.

jeppe-style replied:
see https://github.com/benchflow/benchflow/pull/465#discussion_r124842226

VincenzoFerme commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/configuration/goal/explorationspace/workload/users/UserValuesYamlProtocol.scala:
```diff
@@ -0,0 +1,162 @@
+  }
+
+  private def generateStepList(range: List[Int], step: Int, prefix: String): Try[List[Int]] = prefix match {
+    case "+" =>
```
All the checks are generic and can go in the superclass, only the actual function is specific but it is always the same given the type and can be called by the supertype.

jeppe-style replied:
see https://github.com/benchflow/benchflow/pull/465#discussion_r124842226

VincenzoFerme commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/configuration/goal/explorationspace/workload/users/UserValuesYamlProtocol.scala:
```diff
@@ -0,0 +1,162 @@
+ * @author Jesper Findahl (jesper.findahl@gmail.com)
+ *         created on 2017-05-23
+ */
+object UserValuesYamlProtocol extends DefaultYamlProtocol {
```
You deleted https://github.com/benchflow/benchflow/pull/465/files#diff-5b7b9c50538d06a01fd7c3dca9fa2dd4 and added this class. This class has much code that is similar to other classes (or it is going to be similar).

According to me we should:

- Have base types (equivalent to primitive types in a programming language), that are defined independently of the DSL entity so that can be reused to obtain values. In this case `Int` would be an independent type, and `User` would use it. This way you can use the same definition to get the value of `users` independently of where in the DSL it is defined (since we allow it to be defined as exploration parameter or as configuration. 
- Have a generic definition for shared elements, as `values`, `range`, `step` that can be reused for all the elements needing it. This generic definition is reused, and according to the context (e.g., reading the `users` or the `memory` definition) the correct Reader can be called. 


Update this PR or open an issue to describe the needed changes.

jeppe-style replied:
I did not do that because for each type the allowed ranges might be different. I think it is better to do this refactoring once there is a need. I need now to focus on finishing for the demo.

jeppe-style committed:
Adds DSL-key to custom exception
Clarifications in UserValuesYamlProtocol

VincenzoFerme commented:
I moved in https://github.com/benchflow/benchflow/issues/397#issuecomment-314020445, what can be implemented when extending the support to other types, what is left for this pull request it:

- https://github.com/benchflow/benchflow/pull/465#discussion_r123870137
- The same kind of checks should happen all in the same place https://github.com/benchflow/benchflow/pull/465#discussion_r124005724, https://github.com/benchflow/benchflow/pull/465#discussion_r124006556
- Check that this is done for all the operators https://github.com/benchflow/benchflow/pull/465#discussion_r124008640

VincenzoFerme committed:
Refactors and Injects Behaviour with Traits

It also use traits to make the validation and checks more reusable across different types.

VincenzoFerme committed:
Adds new Tests and Update old Ones after Refactoring

VincenzoFerme commented:
@jeppe-style I implemented https://github.com/benchflow/benchflow/pull/465#issuecomment-314020453. Can you please review the pull request for the changes I have made.

jeppe-style commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/yamlprotocol/valuesrange/IntValuesRangeYamlProtocol.scala:
```diff
@@ -0,0 +1,103 @@
+        Success(tryStepList.get)
+      else
+        Failure(tryStepList.failed.get)
+
```
I think in these lines we can directly return `generateStepList` as it has the same return type

jeppe-style commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/yamlprotocol/valuesrange/IntValuesRangeYamlProtocol.scala:
```diff
@@ -0,0 +1,103 @@
+  protected def getIntStep(step: String) = {
+
+    val (prefix: String, triedStepValue: Try[Int]) = tryGetStep(step)
+
```
It seems this method is only called once so we could integrate it into this method, also because we explicitly get an int step in the `tryGetStep` method

jeppe-style commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/yamlprotocol/valuesrange/IntValuesRangeYamlProtocol.scala:
```diff
@@ -0,0 +1,103 @@
+      if(stepValue == 0) {
+        Failure(BenchFlowDeserializationExceptionMessage("step function cannot be +0"))
+      } else if (increasingRange.isFailure) {
+        Failure(increasingRange.failed.get)
```
we can directly return the `increasingRange` as it is already a `Failure`

jeppe-style commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/yamlprotocol/valuesrange/IntValuesRangeYamlProtocol.scala:
```diff
@@ -0,0 +1,103 @@
+
+    case "+" =>
+      val increasingRange = assertIncreasingRange(range)
+
```
The value inside the `Try` is never used

jeppe-style commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/yamlprotocol/valuesrange/IntValuesRangeYamlProtocol.scala:
```diff
@@ -0,0 +1,103 @@
+      } else if (range.head == 0) {
+        Failure(BenchFlowDeserializationExceptionMessage("first element in the list cannot be *0"))
+      } else if (increasingRange.isFailure) {
+        Failure(increasingRange.failed.get)
```
we can directly return the `increasingRange` as it is already a `Failure`

jeppe-style commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/yamlprotocol/valuesrange/IntValuesRangeYamlProtocol.scala:
```diff
@@ -0,0 +1,103 @@
+        RangeWithStep.stepList(range, stepValue, (a: Int, b: Int) => a + b)
+
+    case "*" =>
+      val increasingRange = assertIncreasingRange(range)
```
The value inside the `Try` is never used

jeppe-style commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/yamlprotocol/valuesrange/IntValuesRangeYamlProtocol.scala:
```diff
@@ -0,0 +1,103 @@
+      }
+
+    case "^" =>
+      val increasingRange = assertIncreasingRange(range)
```
The value inside the `Try` is never used

jeppe-style commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/yamlprotocol/valuesrange/IntValuesRangeYamlProtocol.scala:
```diff
@@ -0,0 +1,103 @@
+      } else if (range.head == 0 || range.head == 1) {
+        Failure(BenchFlowDeserializationExceptionMessage("first element in the list cannot be ^0 or ^1"))
+      } else if (increasingRange.isFailure) {
+        Failure(increasingRange.failed.get)
```
we can directly return the `increasingRange` as it is already a `Failure`

jeppe-style commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/yamlprotocol/valuesrange/MonotonicPositiveIntValuesRangeYamlProtocol.scala:
```diff
@@ -0,0 +1,77 @@
+    val rangeValid = assertRangeIsValid(range)
+
+    if (rangeValid.isFailure) {
+      Failure(rangeValid.failed.get)
```
we can directly return the `rangeValid` as it is already a `Failure`

jeppe-style commented on benchflow-dsl/src/main/scala/cloud/benchflow/dsl/definition/yamlprotocol/valuesrange/MonotonicPositiveIntValuesRangeYamlProtocol.scala:
```diff
@@ -0,0 +1,77 @@
+        Success(range)
+      }
+    } else {
+      Failure(rangeValid.failed.get)
```
we can directly return the `rangeValid` as it is already a Failure

jeppe-style reviewed (commented):
I had a quick look now and added some comments. To me it seems there is some repetition also in some places since there is some nesting in the traits.

VincenzoFerme commented:
@jeppe-style I committed an updated version where I apply your comments. We can deal with repetitions when implementing https://github.com/benchflow/benchflow/issues/397#issuecomment-314408192

jeppe-style commented:
@VincenzoFerme ok for me. I cannot approve since I'm the original author.
